keps/sig-node/753-sidecar-containers/README.md

```diff
@@ -843,24 +992,57 @@ examples:
 
 ### Termination of containers
 
-In alpha sidecar containers will be terminated as regular containers. No special
+In Alpha sidecar containers will be terminated as regular containers. No special
 or additional signals will be supported.
 
-We will collect feedback from alpha implementation and plan to improve
-termination in Beta. When Pods with sidecars are terminated:
+In Beta we have thought about introducing additional termination grace period fields
+to manage termination duration
+([draft proposal](https://docs.google.com/document/d/1B01EdgWJAfkT3l6CIwNwoskQ71eyuwet3mjiQrMQbU8))
+and leverage these fields to add reverse order termination of sidecar containers
+after the primary containers terminate.
 
+However, we decided on an alternative that doesn't require additional fields or hooks while keeping
+the desired behaviors when Pods with sidecars are terminated. While original approach works better
+with truly graceful termination where consistency is more important than time taken, proposed approach
+works for that scenario as well as a more and more popular scenario of limited time to terminate when
+graceful termination is set by external requirement and Pods needs to do best to gracefully terminate
+as much as possible (think of a Spot Instances with 30 seconds notification).
+
+Here is the proposed approach:
+1. Sidecar containers that have a `PreStop` hook will be notified when the Pod has begun terminating
```

Comment:
I want to float for consideration: do we want a different hook?  We could define `preStopPod` with clear semantics that are meaningful.  If we go beta with `preStop` the way this is specc'ed, and we're wrong, we have to add `preStopContainer` or something which is less obvious.  

The `preStop` docs also say that it is called in case of a livenessProbe failure.

Reply:
liveness probe is a very good point. I totally forgot about this.

So if people implement `preStop` as a "signal" and initiate flushing of buffers, while in no-buffer mode, than buffers will be lost.

If sidecar is implemented as full flush in preStop, while in no-buffer mode, than liveness behavior will be OK. The problem will be that the sidecar is potentially delaying the Pod termination for the duration of it's preStop.

So if buffers are less important on Pod termination while they are very important on liveness probe failure, than we have a problem.

Reply:
I am hesitant to add a new hook that only has meaning for the sidecars. Do you think `preStopPod ` may be useful for regular containers?

Reply:
I think the fact that `livenessProbe` uses `preStop` _requires_ a different hook.

Consider a logging sidecar.
* it starts up
* call `postStart`
* app containers start writing logs
* logger fails livenessProbe
* call `preStop` (which, now that I think of it, is kind of weird)
* restart logger
* Pod is deleted
* call `preStop`
* app container stays alive until SIGKILL
* logger is killed with 0 grace

Where in there should it have started aggressively flushing buffers?

I went back to the diagrams I started earlier this year and realized I don't think I shared them publicly: https://docs.google.com/presentation/d/1e-qJWe6He2qjt0PFiZtocqP8VbkrwKAI9B4X72L0bGk/edit?usp=sharing&resourcekey=0-N77W7Q5UHuN5dFtekjmAIA - I have not added sidecars here, but looking at this, `preStop` is pretty clearly ambiguous.
 
If we can't change the definition of `preStop`, we need something like `preStopPod` or `podDeleted`.  Adding API at this point means either a second feature gate or this stays in Alpha.

Thought experiment: if we consider SIGTERM as just another form of lifecycle hook, what would you name it?  `stop` ?  That's a little confusing - one might assume that `preStop` is always followed by `stop`, but it's not.  Maybe `eol` or `terminated`?  That way you could say that a pod being deleted will see:

* deletionTimestamp is set
* all containers get `podDeleted` hook
* app containers get `preStop`
* app containers get `terminated` (SIGTERM)
* wait up to TGPS
* app containers get `SIGKILL`
* if there is time left
  * sidecars get `preStop`
  * sidecars get `terminated` (SIGTERM)
  * wait up to remainder of TGPS
  * sidecars get SIGKILL

In fact we could literally model `terminated` as a hook in `lifecycle`, which defaults to `signal: { name: "SIGTERM" }` (solving a different long-standing issue :) -- but not in this KEP.

Unresolved: do we send preStop to all sidecars at once or one at a time in reverse order?  If we were to later add a per-container grace period, I think we would say one at a time, right?

Reply:
In my mind it's simpler:

- `preStop` == prepare to be terminated. You may still be needed though.
- `SIGTERM` == we strongly insist you exit on your own, it's time now and nobody needs you.
- `SIGKILL` == ok, we warned you, done.

In this sense, liveness is not much different than pod completion. On `preStop` you start flushing aggressively, but can still accept logs. Prepared to be deleted. If you want to make liveness `preStop` to be faster, you can always set `terminationGracePeriodSeconds` override on the liveness probe.

Thinking more about it, new lifecycle hook will just complicate things

Reply:
`preStop` on a liveness failure is a totally different thing from a deleted pod.  There's no need to flush buffers on a liveness failure.  In fact, I think you DON'T WANT TO - you want to do as little work as possible and die, so you can be restarted, right?

If I have a pod who fails liveness, we call preStop.  Do we also send SIGTERM, wait, SIGKILL?  I don't think so (except the new probe-level setting, which I hate).  The docs make it sound like we do, but I didn't see SIGTERM in that case and my diagrams do not have it.

If we overload preStop this way AND go beta, we can't easily take it back (not impossible but hard).    Are you willing to put this all behind a new alpha gate?

Reply:
I think diagram is wrong. We are calling a regular `killContainer` (https://github.com/kubernetes/kubernetes/blob/622509830c1038535e539f7d364f5cd7c3b38791/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L1093).

> `preStop` on a liveness failure is a totally different thing from a deleted pod. There's no need to flush buffers on a liveness failure. In fact, I think you DON'T WANT TO - you want to do as little work as possible and die, so you can be restarted, right?

This is exactly the reason the override was introduced. So liveness probe's preStop can be cut short.

Depending on how container designed, you may expect that the buffers need flushing after liveness probe failed. But in general I agree this may not be very reliable.

> If I have a pod who fails liveness, we call preStop. Do we also send SIGTERM, wait, SIGKILL? I don't think so (except the new probe-level setting, which I hate). The docs make it sound like we do, but I didn't see SIGTERM in that case and my diagrams do not have it.

Yes, those will be sent.

> If we overload preStop this way AND go beta, we can't easily take it back (not impossible but hard). Are you willing to put this all behind a new alpha gate?

The proposal is to keep the behavior that exists today.

I agree with your reasoning. It feel more natural to have a signal differentiated. But it is not a problem scoped to sidecars, all containers needs changing if we decide to improve this. One way to do it in future may be to pass additional arguments to the lifecycle hook to distinguish signals. Fixing is one way to sidecars and another way to regular containers will be the worst of both worlds

Reply:
> I think diagram is wrong.

I went and proved you right.  I updated the drawings.

> This is exactly the reason the override was introduced. So liveness probe's preStop can be cut short.

We never should have called preStop on liveness fail.  It should have been different.

> The proposal is to keep the behavior that exists today.

I barely count alpha as a commitment.  Once we do this in beta, it's a much stronger statement of "we considered it and this is how we want it".  What that means is `preStop` == `podDelete`.

So rewriting my timeline:

* `deletionTimestamp` is set
* all containers get `preStop` hook
* app containers get `terminate` (implicitly == SIGTERM)
* wait up to TGPS
* app containers get SIGKILL
* iff there is time left
  * sidecars get `terminate` (implicitly == SIGTERM)
  * wait up to remainder of TGPS
  * sidecars get SIGKILL

> It feel more natural to have a signal differentiated. But it is not a problem scoped to sidecars, all containers needs changing if we decide to improve this. 

I don't see what you mean?  If we spec `preStop` as "this container is being stopped" and `podDelete` as "the pod is being stopped" (with the guarantees that `podDelete` does not happen in case of liveness fail and `podDelete` always happens before `preStop` in case of pod termination), the only containers who need to adapt are sidecars, and they already need to adapt.

What's the reason to not add a specific `podDelete` hook?  Is it just to avoid another gate and/or alpha (which is somewhat valid)?  Or is there actually a semantic argument?

In your model: do we send `preStop` to all sidecars at once or one at a time in reverse order? If we were to later add a per-container grace period, I think we would say one at a time, right?

Reply:
> > The proposal is to keep the behavior that exists today.
> 
> I barely count alpha as a commitment. Once we do this in beta, it's a much stronger statement of "we considered it and this is how we want it". What that means is `preStop` == `podDelete`.

What I meant here is that we should be keeping the behavior of calling `preStop` on liveness probe failure for sidecars similarly how it is done on regular container is what I mean by existing behavior. Changing this semantic should ideally be done consistently across regular and sidecar containers.

> > It feel more natural to have a signal differentiated. But it is not a problem scoped to sidecars, all containers needs changing if we decide to improve this.
> 
> I don't see what you mean? If we spec `preStop` as "this container is being stopped" and `podDelete` as "the pod is being stopped" (with the guarantees that `podDelete` does not happen in case of liveness fail and `podDelete` always happens before `preStop` in case of pod termination), the only containers who need to adapt are sidecars, and they already need to adapt.

I read the semantic of `preStop` as a signal to start packing. Container will be stopped soon, but we will do our best to wait for you "pack your things". Not terminating your yet, so please keep the readiness probe up if you can serve things. But letting you know.

> What's the reason to not add a specific `podDelete` hook? Is it just to avoid another gate and/or alpha (which is somewhat valid)? Or is there actually a semantic argument?

For me it is a semantic argument. I want to keep sidecars behavior aligned with other containers in terms of lifecycle hooks. And I think sidecars fit perfectly into the model today to enable most of termination scenarios. 

> In your model: do we send `preStop` to all sidecars at once or one at a time in reverse order? If we were to later add a per-container grace period, I think we would say one at a time, right?

All sidecars at once. I do not have any ordering preferences. @tzneal implementaiton-wise is there any ordering for the regular containers today? I believe it will be sequencial, but non-blocking, correct?

Reply:
> What I meant here is that we should be keeping the behavior of calling preStop on liveness probe failure for sidecars similarly how it is done on regular container

Oh.  I agree with that.  IMO that mnakes it more obvious that `preStop` and `podDelete` are different things.  I don't think we can change it.

> I read the semantic of preStop as a signal to start packing. Container will be stopped soon

Yes, I think that's true in either approach.

>> What's the reason to not add a specific podDelete hook? Is it just to avoid another gate and/or alpha (which is somewhat valid)? Or is there actually a semantic argument?
> 
> For me it is a semantic argument. I want to keep sidecars behavior aligned with other containers in terms of lifecycle hooks. And I think sidecars fit perfectly into the model today to enable most of termination scenarios.

It fits fine, if you squint a little.  You say "aligned with other containers " but you are pinning that on one interpretation.  

Today, `preStop` means:
* the container is going down soon (e.g. liveness)
* the pod is going down soon (e.g. deleted)
* start your end-of-life

This is fine because we have no sequence.  As soon as we add a sequence, we need to decide which of those `preStop` means.  Because of the liveness probe semantic, it CAN'T mean "the pod is going down soon".   You are saying it should mean "start your end-of-life" and pods don't get to know why.  I am saying it should mean "the container is going down soon" and a different signal should mean "the pod is going down soon".  Both work, I think.  Your path is simpler and informative.  Other path is richer but more complex.

Is that a fair breakdown?  

>> In your model: do we send preStop to all sidecars at once or one at a time in reverse order? If we were to later add a per-container grace period, I think we would say one at a time, right?
>
> All sidecars at once. I do not have any ordering preferences. @tzneal implementaiton-wise is there any ordering for the regular containers today? I believe it will be sequencial, but non-blocking, correct?

Can we spec that preStop and SIGTERM are always delivered in reverse-startup order, though there may be almost zero time between them?  That leaves room for adding grace periods later if we need.

Reply:
> This is fine because we have no sequence. As soon as we add a sequence, we need to decide which of those `preStop` means. Because of the liveness probe semantic, it CAN'T mean "the pod is going down soon". You are saying it should mean "start your end-of-life" and pods don't get to know why. I am saying it should mean "the container is going down soon" and a different signal should mean "the pod is going down soon". Both work, I think. Your path is simpler and informative. Other path is richer but more complex.
> 
> Is that a fair breakdown?

Yes, it is fair. More signal is better for more advanced scenarios. But complicates documentation and implementation. 

Also we may enrich things by passing more arguments to the lifecycle hooks in future in form of env vars or http headers. It may be better than adding more hooks or change the pod specs.

> Can we spec that preStop and SIGTERM are always delivered in reverse-startup order, though there may be almost zero time between them? That leaves room for adding grace periods later if we need.

Yes, this is a great idea. If we have a definitive order for the regular containers we can spec it as well.

Reply:
> All sidecars at once. I do not have any ordering preferences. @tzneal implementaiton-wise is there any ordering for the regular containers today? I believe it will be sequential, but non-blocking, correct?

No, there's no ordering for regular containers. It's a go-routine to kill each container, and waiting on them all to finish.  https://github.com/kubernetes/kubernetes/blob/6a84edb2cee68fc60d3f4bd15eb461a184303236/pkg/kubelet/kuberuntime/kuberuntime_container.go#L768

Reply:
I don't really care about app containers - more concerned about sidecars.